File: integration-tests/TestApp/app/src/androidTestApi9/java/android/databinding/testapp/SafeUnboxingTest.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package android.databinding.testapp;


import android.databinding.testapp.databinding.SafeUnboxingBinding;
import android.support.test.runner.AndroidJUnit4;
import android.test.UiThreadTest;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.CoreMatchers.is;

public class SafeUnboxingTest extends BaseDataBinderTest<SafeUnboxingBinding> {
    public SafeUnboxingTest() {
        super(SafeUnboxingBinding.class);
    }

    @Override
    protected void setUp() throws Exception {
        super.setUp();
        initBinder();
    }

    @UiThreadTest
    public void testBitShift() {
        mBinder.executePendingBindings();
        assertThat(mBinder.bitShift.intField, is(0));
        mBinder.setAnInt(3);
        mBinder.executePendingBindings();
        assertThat(mBinder.bitShift.intField, is(3 << 3));
    }

    @UiThreadTest
    public void testArrayAccess() {
        mBinder.executePendingBindings();
        assertThat(mBinder.arrayAccess.intField, is(0));

        mBinder.setAnInt(3);
        mBinder.executePendingBindings();
        assertThat(mBinder.arrayAccess.intField, is(0));

        mBinder.setAnArray(new int[]{2,3,4,5});
        mBinder.executePendingBindings();
        assertThat(mBinder.arrayAccess.intField, is(5));

        mBinder.setAnInt(7);
        mBinder.executePendingBindings();
        assertThat(mBinder.arrayAccess.intField, is(0));

        mBinder.setAnInt(null);
        mBinder.executePendingBindings();
        assertThat(mBinder.arrayAccess.intField, is(2));
    }

    @UiThreadTest
    public void testIntEq() {
        mBinder.executePendingBindings();
        assertThat(mBinder.ternary.intField, is(2));
        mBinder.setAnInt(1);
        mBinder.executePendingBindings();
        assertThat(mBinder.ternary.intField, is(3));
    }

    @UiThreadTest
    public void testBooleanTernary() {
        mBinder.executePendingBindings();
        assertThat(mBinder.ternary2.intField, is(3));
        mBinder.setABoolean(true);
        mBinder.executePendingBindings();
        assertThat(mBinder.ternary2.intField, is(2));
    }

    @UiThreadTest
    public void testSetter1() {
        mBinder.executePendingBindings();
        assertThat(mBinder.setter1.intField, is(0));
        assertThat(mBinder.setter1.integerField, is(nullValue()));
        mBinder.setAnInt(3);
        mBinder.executePendingBindings();
        assertThat(mBinder.setter1.intField, is(3));
        assertThat(mBinder.setter1.integerField, is(3));
    }

    @UiThreadTest
    public void testAdapter() {
        mBinder.executePendingBindings();
        assertThat(mBinder.checkBox.isChecked(), is(false));
        mBinder.setABoolean(true);
        mBinder.executePendingBindings();
        assertThat(mBinder.checkBox.isChecked(), is(true));
    }
}

File: compilationTests/src/test/java/android/databinding/compilationTest/SafeUnboxingTest.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.databinding.compilationTest

import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

@RunWith(Parameterized::class)
class SafeUnboxingTest(pair: Pair<String, Boolean>) : BaseCompilationTest(true) {
    private val expr  = pair.first
    private val shouldWarn = pair.second
    companion object {
        @JvmStatic
        @Parameterized.Parameters
        fun getParams() =
            listOf(
                    "@{myInt > 0 ? `t` : `f`}" to true,
                    "@{myInt == null ? `t` : `f`}" to false,
                    "@{myInt != null ? `t` : `f`}" to false,
                    "@{myInt ?? 3}" to false,
                    "@{myInt ?? Integer.valueOf(`3`)}" to true,
                    "@{myInt ?? Integer.parseInt(`3`)}" to false,
                    "@{myInt == null ? Integer.valueOf(`3`) : myInt}" to true,
                    "@{null == myInt ? `t` : `f`}" to false,
                    "@{null != myInt ? `t` : `f`}" to false,
                    "@{null == myInt ? Integer.valueOf(`3`) : myInt}" to true
            )
    }
    @Test
    fun check() {
        prepareProject()
        copyResourceTo("/layout/safe_unbox_layout.xml",
                "/app/src/main/res/layout/main.xml",
                mapOf("EXPR" to expr))
        val result = runGradle("assembleDebug")
        assertThat(result.error, result.resultCode, `is`(0))

        assertThat(result.error, result.bindingWarnings.any {
            it.contains(" is a boxed field but needs to be un-boxed to execute")
        }, `is`(shouldWarn))
    }
}